src/qspi.rs

```diff
@@ -0,0 +1,783 @@
+//! QSPI interface
+//!
+//! This module implements the QuadSPI interface which allows high speed
+//! communication with external flash memory.
+//!
+//! Limitations:
+//!     - Interrupts are not supported.
+//!     - Status polling mode is not supported.
+
+// Based on work by Unizippro <madas@astrupa.dk> in stm32l4xx-hal.
+
+use core::{any::TypeId, marker::PhantomData};
+
+use crate::{
+    gpio::alt::{self, QuadSpiBank},
+    pac::QUADSPI,
+    rcc::Enable,
+};
+
+pub trait QspiPins {
+    type Pins;
+}
+
+impl<T> QspiPins for T
+where
+    T: QuadSpiBank,
+{
+    type Pins = (T::Ncs, T::Io0, T::Io1, T::Io2, T::Io3, alt::quadspi::Clk);
+}
+
+pub struct DualFlash;
+
+impl QspiPins for DualFlash {
+    type Pins = (
+        alt::quadspi::Bk1Ncs,
+        alt::quadspi::Bk2Ncs,
+        alt::quadspi::Bk1Io0,
+        alt::quadspi::Bk1Io1,
+        alt::quadspi::Bk1Io2,
+        alt::quadspi::Bk1Io3,
+        alt::quadspi::Bk2Io0,
+        alt::quadspi::Bk2Io1,
+        alt::quadspi::Bk2Io2,
+        alt::quadspi::Bk2Io3,
+        alt::quadspi::Clk,
+    );
+}
+
+pub struct Qspi<Pins: QspiPins> {
+    qspi: QUADSPI,
+    config: QspiConfig,
+    _marker: PhantomData<Pins>,
+}
```

Comment:
I'd rename `Pins` -> `BANK` and keep pins inside. This allows you to add `release` for `Qspi`.
```suggestion
pub struct Qspi<BANK: QspiPins> {
    qspi: QUADSPI,
    config: QspiConfig,
    pins: BANK::Pins,
}
```